### ISSUE TITLE
`bos_token` used to mask instead of `boi_token`

### ISSUE THREAD

**Issue by Vidit-Ostwal** [ariG23498/gemma3-object-detection#35]
Hi @sergiopaniego, @ariG23498 

I was trying to understand how the data loader works, and typically what exact input is going in to the model!

I came across this particular line of code.

Consider the case when 
`cfg.model_id = 'google/gemma-3-4b-pt'`

In this case `tokenizer.boi_token` which is `'<start_of_image>'` is being used 
https://github.com/ariG23498/gemma3-object-detection/blob/2413e6a126ff68e14bafb1d56c9f99d5d164d726/utils.py#L85-L92

while trying to find the `image_token_id` `bos_token` is used.
https://github.com/ariG23498/gemma3-object-detection/blob/2413e6a126ff68e14bafb1d56c9f99d5d164d726/utils.py#L100-L108

I think this should be `boi_token` to properly find the image_token_id.

I also went to official website of Gemma3 fine tuned with Qlora
https://ai.google.dev/gemma/docs/core/huggingface_vision_finetune_qlora
To check how they have suggested to fine-tune.
They have also used `boi_token`

from trl import SFTConfig
```python
args = SFTConfig(
    output_dir="gemma-product-description",     # directory to save and repository id
    num_train_epochs=1,                         # number of training epochs
    per_device_train_batch_size=1,              # batch size per device during training
    gradient_accumulation_steps=4,              # number of steps before performing a backward/update pass
    gradient_checkpointing=True,                # use gradient checkpointing to save memory
    optim="adamw_torch_fused",                  # use fused adamw optimizer
    logging_steps=5,                            # log every 5 steps
    save_strategy="epoch",                      # save checkpoint every epoch
    learning_rate=2e-4,                         # learning rate, based on QLoRA paper
    bf16=True,                                  # use bfloat16 precision
    max_grad_norm=0.3,                          # max gradient norm based on QLoRA paper
    warmup_ratio=0.03,                          # warmup ratio based on QLoRA paper
    lr_scheduler_type="constant",               # use constant learning rate scheduler
    push_to_hub=True,                           # push model to hub
    report_to="tensorboard",                    # report metrics to tensorboard
    gradient_checkpointing_kwargs={
        "use_reentrant": False
    },  # use reentrant checkpointing
    dataset_text_field="",                      # need a dummy field for collator
    dataset_kwargs={"skip_prepare_dataset": True},  # important for collator
)
args.remove_unused_columns = False # important for collator

# Create a data collator to encode text and image pairs
def collate_fn(examples):
    texts = []
    images = []
    for example in examples:
        image_inputs = process_vision_info(example["messages"])
        text = processor.apply_chat_template(
            example["messages"], add_generation_prompt=False, tokenize=False
        )
        texts.append(text.strip())
        images.append(image_inputs)

    # Tokenize the texts and process the images
    batch = processor(text=texts, images=images, return_tensors="pt", padding=True)

    # The labels are the input_ids, and we mask the padding tokens and image tokens in the loss computation
    labels = batch["input_ids"].clone()

    # Mask image tokens
    image_token_id = [
        processor.tokenizer.convert_tokens_to_ids(
            processor.tokenizer.special_tokens_map["boi_token"]
        )
    ]
    # Mask tokens for not being used in the loss computation
    labels[labels == processor.tokenizer.pad_token_id] = -100
    labels[labels == image_token_id] = -100
    labels[labels == 262144] = -100

    batch["labels"] = labels
    return batch
```

**Comment by Vidit-Ostwal** [comment]
Let me know if I am missing something?

I am not sure how much effect it has if we don't mask one particular token?
Could you guide me here a bit.

**Comment by Vidit-Ostwal** [comment]
Also any reason why we didn't used chat templates?
